text/3127-trim-path.md

```diff
@@ -0,0 +1,192 @@
+
+`trim-path` is a profile setting which can be set to either `true` or `false`. This is enabled by default when you do a release build,
+such as via `cargo build --release`. You can also manually override it by specifying this option in `Cargo.toml`:
+```toml
+[profile.dev]
+trim-path = true
+
+[profile.release]
+trim-path = false
+```
+
+With `trim-path` option enabled, the compilation process will not introduce any absolute paths into the build output. Instead, paths containing
+certain prefixes will be replaced with something stable by the following rules:
+
+1. Path to the source files of the standard and core library will begin with `/rustc/[rustc version]`.
+   E.g. `/home/username/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs` -> 
+   `/rustc/1.52.1/library/core/src/result.rs`
+2. Path to the working directory will be replaced with `.`. E.g. `/home/username/crate/src/lib.rs` -> `./src/lib.rs`.
+3. Path to packages outside of the working directory will be replaced with `[package name]-[version]`. E.g. `/home/username/deps/foo/src/lib.rs` -> `foo-0.1.0/src/lib.rs`
+
+If using MSVC toolchain, path to the .pdb file containing debug information are be embedded as the file name of the .pdb file only, wihtout any path
+information.
+
+With `trim-path` option disabled, the embedding of path to the source files of the standard and core library will depend on if `rust-src` component is present. If it is, then the real path pointing to a copy of the source files on your file system will be embedded; if it isn't, then they will
+show up as `/rustc/[rustc version]/library/...` (just like when `trim-path` is enabled). Path to all other source files will not be affected.
+
+Note that this will not affect any hard-coded paths in the source code.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `trim-path` implementation in Cargo
+We only need to change the behaviour for `Test` and `Build` compile modes. 
+
+If `trim-path` is enabled, Cargo will emit two `--remap-path-prefix` arguments to `rustc` for each compilation unit. One mapping is from the path of 
+the local sysroot to `/rustc/[rust version]`. The other mapping depends on if the package containing the compilation unit is under the working
+directory. If it is, then the mapping is from the absolute path to the working directory to `.`. If it's outside the working directory, then the
+mapping is from the absolute path of the package root to `[package name]-[package version]`.
+
+Some interactions with compiler-intrinstic macros need to be considered, though these are entirely down to `rustc`'s implementation of
+`--remap-path-prefix`:
+1. Path (of the current file) introduced by [`file!()`](https://doc.rust-lang.org/std/macro.file.html) *will* be remapped. **Things may break** if
+   the code interacts with its own source file at runtime by using this macro.
```

Comment:
If you want an example of things breaking to link to: <https://github.com/rust-num/num-traits/issues/139>

Reply:
We could just not apply remapping when building `build.rs`, since it'll be up to the build script to maintain privacy and reproducibility anyway

Reply:
If `file!()` is relative to `CARGO_MANIFEST_DIR`, then `build.rs` should be able to fix these paths, and it's not even difficult: `manifest_dir.join(relative_or_absolute_file_path)` just works.

Reply:
The fix in `num` was to just emit a relative path, so I assume Cargo already treats relative paths as relative to `CARGO_MANIFEST_DIR` (though I see that isn't documented on <https://doc.rust-lang.org/nightly/cargo/reference/build-scripts.html#rerun-if-changed>).